monkey/monkey_island/cc/repository/mongo_machine_repository.py

```diff
@@ -0,0 +1,102 @@
+from ipaddress import IPv4Address
+from threading import Lock
+from typing import Any, MutableMapping, Sequence
+
+from pymongo import MongoClient
+
+from common.types import HardwareID
+from monkey_island.cc.models import Machine, MachineID
+
+from . import IMachineRepository, RemovalError, RetrievalError, StorageError, UnknownRecordError
+from .consts import MONGO_OBJECT_ID_KEY
+
+
+class MongoMachineRepository(IMachineRepository):
+    """A repository used to store and retrieve Machines in MongoDB"""
+
+    def __init__(self, mongo_client: MongoClient):
+        self._machines_collection = mongo_client.monkey_island.machines
+        self._id_lock = Lock()
+        self._next_id = self._get_biggest_id()
+
+    def _get_biggest_id(self) -> MachineID:
+        try:
+            return self._machines_collection.find().sort("id", -1).limit(1)[0]["id"]
+        except IndexError:
+            return 0
+
+    def create_machine(self) -> Machine:
+        try:
+            next_id = self._get_next_id()
+            new_machine = Machine(id=next_id)
+            self._machines_collection.insert_one(new_machine.dict(simplify=True))
+
+            return new_machine
+        except Exception as err:
+            raise StorageError(f"Error creating a new machine: {err}")
+
+    def _get_next_id(self) -> MachineID:
+        with self._id_lock:
+            self._next_id += 1
+            return self._next_id
+
+    def update_machine(self, machine: Machine):
+        try:
+            result = self._machines_collection.replace_one(
```

Comment:
Consider using `upsert=True`. This would create a machine if it didn't exist. But then we need to modify the machine object by reference or return a new one

Reply:
See https://github.com/guardicore/monkey/pull/2228#discussion_r959646019